src/canvas/PatternPiece.tsx

```diff
@@ -0,0 +1,97 @@
+import { PatternPath } from './PatternPaths/PatternPath';
+import { Vector } from 'canvas/Geometry/Vector';
+import { LineSegment } from './Geometry/LineSegment';
+import { PathIntersection } from './PathIntersection';
+
+export class PatternPiece {
+    private _paths: PatternPath[];
+    private _allowancePaths: PatternPath[];
+
+    /**
+     * Constructs a PatternPiece object from the provided 
+     * PatternPath array.
+     * 
+     * @param paths ordered paths going counterclockwise 
+     *              around the piece. The end of a path should
+     *              be the exact same point as the start of the
+     *              following path
+     */
+    constructor(paths: PatternPath[]) {
+        this._paths = paths;
+        this._allowancePaths = this._computeAllowancePaths();
+    }
+    
+    /**
+     * Returns all of the PatternPath edges and PatternPath 
+     * allowances of the pattern piece
+     */
+    getAllPaths = (): PatternPath[] => {
+        return [...this._paths, ...this._allowancePaths];
+    };
+
+    /**
+     * Translates the patternPiece in the plane following the
+     * provided displacement vector.
+     * 
+     * @param displacement 
+     */
+    translate = (displacement: Vector): void => {
+        this._paths.forEach(path => {
+            path.translate(displacement);
+        });
+        this._allowancePaths.forEach(path => {
+            path.translate(displacement);
+        });
+    };
+
+    private _computeAllowancePaths = (): PatternPath[] => {
+        // First, compute the allowance paths.
+        // These are prolonged paths that run parallel to the edges.
+        const allowances = this._paths.map(path => path.getAllowance());
+
+        // Then, edge by edge, find the intersection between the alloance of 
+        // an edge and the allowance of the next and shorten the allowances
+        // accordingly
+        const numAllowances = allowances.length;
+        for (let i = 0; i < numAllowances; i++) {
+            const intersections = this._findIntersectionBetweenPaths(allowances[i], allowances[(i + 1) % numAllowances]);
+            if (intersections) {
+                allowances[i].trimAtPoint(intersections[0].point, intersections[0].indexOfSegmentCrossed, false);
```

Comment:
Yes. It always returns 2. We need both because we need to know the index of the Segment to which the intersection belongs in each Path in order to trim them.

Reply:
Cool, would you add this to the comments please?

Reply:
Yes. I added comments to _findIntersectionBetweenPaths and _computeAllowancePaths in PatternPiece.